include/matx/kernels/resample_poly.cuh

```diff
@@ -217,6 +229,254 @@ __global__ void ResamplePoly1D(OutType output, InType input, FilterType filter,
     }
 }
 
+template <int THREADS, typename FilterType>
+__device__ inline void ResamplePoly1D_LoadFilter(typename FilterType::scalar_type *s_filter, const FilterType &filter)
+{
+    const index_t filter_len = filter.Size(0);
+    const int tid = threadIdx.x;
+    if (filter_len % 2 == 0) {
+        for (int t = tid; t < filter_len; t += THREADS) {
+            s_filter[t+1] = filter.operator()(t);
+        }
+        if (tid == 0) {
+            s_filter[0] = static_cast<typename FilterType::scalar_type>(0);
+        }
+    } else {
+        for (int t = tid; t < filter_len; t += THREADS) {
+            s_filter[t] = filter.operator()(t);
+        }
+    }
+    __syncthreads();    
+}
+
+template <int THREADS, typename OutType, typename InType, typename FilterType, typename index_t>
+__launch_bounds__(MATX_RESAMPLE_POLY_MAX_NUM_THREADS)
+__global__ void ResamplePoly1D_ElemBlock(OutType output, InType input, FilterType filter,
+                    index_t up, index_t down, index_t elems_per_thread)
+{
+    using output_t = typename OutType::scalar_type;
+    using input_t = typename InType::scalar_type;
+    using filter_t = typename FilterType::scalar_type;
+
+    extern __shared__ uint8_t smem_filter[];
+    filter_t *s_filter = reinterpret_cast<filter_t *>(smem_filter);
+
+    constexpr int Rank = OutType::Rank();
+    const index_t output_len = output.Size(Rank-1);
+    index_t filter_len = filter.Size(0);
+    const index_t input_len = input.Size(Rank-1);
+
+    const size_t filter_sz_bytes = (filter_len % 2 == 0) ? sizeof(filter_t)*(filter_len+1) : sizeof(filter_t)*filter_len;
+    const bool load_filter_to_smem = (filter_sz_bytes <= MATX_RESAMPLE_POLY_MAX_SMEM_BYTES);
+
+    const int elem_block = blockIdx.z;
+    const int tid = threadIdx.x;
+    // const int THREADS = blockDim.x;
+
+    if (load_filter_to_smem) {
+        ResamplePoly1D_LoadFilter<THREADS, FilterType>(s_filter, filter);
+        if (filter_len % 2 == 0) {
+            filter_len++;
+        }
+    }
+
+    // All but the last dim are batch indices
+    const int batch_idx = blockIdx.x;
+    auto bdims = BlockToIdx(output, batch_idx, 1);
+
+    // Scale the filter coefficients by up to match scipy's convention
+    const filter_t scale = static_cast<filter_t>(up);
+    const index_t max_input_ind = input_len - 1;
+
+    const index_t filter_len_half = filter_len/2;
+    const index_t filter_central_tap = (filter_len-1)/2;

```

Comment:
Isn't this just filter_len/2 or do you need to be on the left side for even lengths?

Reply:
Yes, this is to handle even-lengths. The logic in the subsequent loops effectively assumes an odd-length filter (and thus a central tap). In the case of storing the filter to shared memory, a zero is prepended to the filter (prior to flipping for convolution) to make the length odd. Thus, for an even-length filter not stored to shared memory, (filter_len-1)/2 always references the same coefficient as filter_len/2 for filters stored to shared memory. I had to double-check, so I will update the PR to include a comment here.